File: problems/graphs/detect_arbitrage/python/detect_arbitrage.py
import math


def detectArbitrage(exchangeRates: list[list[float]]) -> bool:
    logExchangeRates: list[list[float]] = [
        [-math.log(rate) for rate in row] for row in exchangeRates
    ]
    n: int = len(logExchangeRates)

    def hasNegativeCycle() -> bool:
        distances: list[int] = [float("inf")] * n
        distances[0] = 0

        for _ in range(n - 1):
            for u in range(n):
                for v in range(n):
                    if distances[u] + logExchangeRates[u][v] < distances[v]:
                        distances[v] = distances[u] + logExchangeRates[u][v]

        for u in range(n):
            for v in range(n):
                if distances[u] + logExchangeRates[u][v] < distances[v]:
                    return True

        return False

    return hasNegativeCycle()

File: problems/graphs/detect_arbitrage/js/detect_arbitrage.js
export const detectArbitrage = (() => {
  const hasNegativeCycle = (exchangeRates) => {
    const distances = Array(exchangeRates.length).fill(Infinity);
    distances[0] = 0;
    const logExchangesRates = exchangeRates.map((row) =>
      row.map((rate) => -Math.log(rate)),
    );

    for (let i = 0; i < exchangeRates.length - 1; i++) {
      for (let u = 0; u < exchangeRates.length; u++) {
        for (let v = 0; v < exchangeRates.length; v++) {
          if (distances[u] + logExchangesRates[u][v] < distances[v]) {
            distances[v] = distances[u] + logExchangesRates[u][v];
          }
        }
      }
    }

    for (let u = 0; u < exchangeRates.length; u++) {
      for (let v = 0; v < exchangeRates.length; v++) {
        if (distances[u] + logExchangesRates[u][v] < distances[v]) {
          return true;
        }
      }
    }
    return false;
  };

  return (exchangeRates) => hasNegativeCycle(exchangeRates);
})();
